kythe/go/util/markedsource/resolve.go

```diff
@@ -0,0 +1,152 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package markedsource
+
+import (
+	"fmt"
+
+	"kythe.io/kythe/go/util/kytheuri"
+	"kythe.io/kythe/go/util/schema/edges"
+	"kythe.io/kythe/go/util/schema/facts"
+
+	"google.golang.org/protobuf/proto"
+
+	cpb "kythe.io/kythe/proto/common_go_proto"
+	spb "kythe.io/kythe/proto/storage_go_proto"
+)
+
+// Resolver produces fully resolved MarkedSources from a set of entries.
+type Resolver struct {
+	nodes map[string]*cpb.MarkedSource
+
+	params  map[string][]string
+	tparams map[string][]string
+	typed   map[string]string
+}
+
+// NewResolver constructs a MarkedSource resolver from the given entries.
+func NewResolver(entries []*spb.Entry) (*Resolver, error) {
+	r := &Resolver{
+		nodes:   make(map[string]*cpb.MarkedSource),
+		params:  make(map[string][]string),
+		tparams: make(map[string][]string),
+		typed:   make(map[string]string),
+	}
+	for _, e := range entries {
+		if e.GetFactName() == facts.Code {
+			ticket := kytheuri.ToString(e.GetSource())
+			var ms cpb.MarkedSource
+			if err := proto.Unmarshal(e.GetFactValue(), &ms); err != nil {
+				return nil, fmt.Errorf("error unmarshalling code for %s: %v", ticket, err)
+			}
+			r.nodes[ticket] = &ms
+		} else if e.GetEdgeKind() != "" {
+			ticket := kytheuri.ToString(e.GetSource())
+			kind, ord, _ := edges.ParseOrdinal(e.GetEdgeKind())
+			if ord < 0 {
+				return nil, fmt.Errorf("invalid ordinal: %d", ord)
+			}
+			switch kind {
+			case edges.Typed:
+				r.typed[ticket] = kytheuri.ToString(e.GetTarget())
+			case edges.Param:
+				params := r.params[ticket]
+				if len(params)-1 < ord {
+					n := make([]string, ord+1)
+					copy(n, params)
+					params = n
+					r.params[ticket] = params
+				}
+				params[ord] = kytheuri.ToString(e.GetTarget())
+			case edges.TParam:
+				tparams := r.tparams[ticket]
+				if len(tparams)-1 < ord {
+					n := make([]string, ord+1)
+					copy(n, tparams)
+					tparams = n
+					r.tparams[ticket] = tparams
+				}
+				tparams[ord] = kytheuri.ToString(e.GetTarget())
+			}
+		}
+	}
+	return r, nil
+}
+
+// Resolve returns the fully resolved MarkedSource for the given source VName.
+// May return nil if no MarkedSource is found.
+func (r *Resolver) Resolve(src *spb.VName) *cpb.MarkedSource {
+	return r.ResolveTicket(kytheuri.ToString(src))
+}
+
+// Resolve returns the fully resolved MarkedSource for the given source ticket.
+// May return nil if no MarkedSource is found.
+func (r *Resolver) ResolveTicket(ticket string) *cpb.MarkedSource {
+	if ticket == "" {
+		return nil
+	}
+	return r.resolve(ticket, r.nodes[ticket])
+}
+
+func (r *Resolver) resolve(ticket string, ms *cpb.MarkedSource) *cpb.MarkedSource {
+	if ms == nil {
+		return ms
+	}
+	switch ms.GetKind() {
+	case cpb.MarkedSource_LOOKUP_BY_PARAM:
+		params := r.params[ticket]
+		if p := params[ms.GetLookupIndex()]; p != "" {
+			return r.ResolveTicket(p)
+		}
```

Comment:
Is failure to find a param when told to do so not an error?

Reply:
That question is part of the reason I made this utility.  It's currently undecided so I just ignored it for now.

Reply:
Can you add a comment mentioning that?